tests/test_notifications.py

```diff
@@ -0,0 +1,53 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import pytest
+from selenium.common.exceptions import TimeoutException
+
+from foxpuppet.browser.notifications import (
+    BaseNotification,
+    AddOnInstallBlocked)
+
+
+@pytest.fixture
+def firefox_options(firefox_options):
+    firefox_options.add_argument('-foreground')
+    return firefox_options
+
+
+@pytest.fixture
+def notification(firefox, webserver, selenium):
+    selenium.get(webserver.url())
+    selenium.find_element_by_link_text('restartless_addon_signed.xpi').click()
+    return firefox.wait_for_notification(AddOnInstallBlocked)
+
+
+def test_open_close_notification(firefox, notification):
+    """Trigger and dismiss a notification"""
+    assert notification is not None
+    notification.close()
+    assert firefox.notification is None
+
+
+@pytest.mark.parametrize('_class, message', [
+    (BaseNotification, 'No notification was shown'),
+    (AddOnInstallBlocked, 'AddOnInstallBlocked was not shown')])
+def test_wait_for_notification_timeout(firefox, _class, message):
+    """Wait for a notification when one is not shown"""
+    with pytest.raises(TimeoutException) as excinfo:
+        firefox.wait_for_notification(_class)
+    assert message in str(excinfo.value)
+
+
+def test_wait_for_no_notification_timeout(firefox, notification):
+    """Wait for no notification when one is shown"""
+    with pytest.raises(TimeoutException) as excinfo:
+        firefox.wait_for_notification(None)
+    assert 'Unexpected notification shown' in str(excinfo.value)
+
+
+def test_notification_with_origin(firefox, webserver, notification):
+    """Trigger a notification with an origin"""
+    assert '{0.host}:{0.port}'.format(webserver) in firefox.notification.origin
```

Comment:
Nice! That formatting (with properties in the brackets)  is new for me. Learned something new.